proposals/2025-02-24_secret_providers.md

```diff
@@ -0,0 +1,173 @@
+
+## Why
+
+The motivation behind this design document is to enhance the security and flexibility of secret management in Prometheus. Currently, Prometheus only supports reading secrets from the filesystem or directly from the configuration file, which can lead to security vulnerabilities and limitations when working with certain service providers.
+
+This proposal introduces secret discovery, similar to service discovery, where different secret providers can contribute code to read secrets from their respective APIs. This would allow for more secure and dynamic secret retrieval, eliminating the need to store secrets in the filesystem and reducing the potential for unauthorized access.
+
+### Pitfalls of the current solution
+
+Storing secrets in the filesystem poses risks, especially in environments like Kubernetes, where any pod on a node can access files mounted on that node. This could expose secrets to attackers. Additionally, configuring secrets through the filesystem often requires extra setup steps in some environments, which can be cumbersome for users.
+
+Storing secrets inline can also pose risks, as the configuration file may still be accessible through the filesystem. Additionally it can lead to configuration files becoming cluttered and difficult to manage.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Allow Prometheus to read secrets remotely from secret providers.
+* Introduce secret discovery, similar to service discovery, where different secret providers can contribute code to read secrets from their respective API.
+
+### Audience
+
+* Prometheus maintainers
+* Alertmanager maintainers
+* Secret providers interested in contributing code
+* Users looking to use secret providers
+
+## Non-Goals
+
+* Support for non-string secret values
+* Secret transformations/processing
+  * Things like String concatenation, base64 encoding/decoding
+* Default values
+
+## How
+
+
+### Configuration
+
+Wherever a secret can be inserted in the configuration file, we will allow users to specify a special YAML tag for secrets.
+
+```
+        password: !secret provider=kubernetes namespace=ns1 secret_id=pass1
+...
+        password: !secret
+          provider: kubernetes
+          namespace: ns1
+          secret_id: pass2
+```
+Additionally there will be a global section to partially configure secret providers to prevent duplication. For instance if the kubernetes provider is used multiple times like above, it can be rewritten as: 
+
+```
+global:
+  base_secrets:
+  - id: myk8secrets
+    provider: kubernetes
+    namespace: ns1
+...
+        password: !secret id=myk8secrets secret_id=pass1
+...
+        password: !secret
+          id: myk8secrets
+          secret_id: pass2
+```
+
+### Inline secrets
+
+When specifying secrets inline in the config file, the inline provider will be used. If a string is passed in, it is automatically converted to an inline provider to be consistent with previous syntax.
```

Comment:
Can you explicitly state what are valid syntaxes users can use for the case a inline secret is used? Also, if they are converted under the hood, that is a implementation detail? do we need to expose in the implementation?

Reply:
You are right, this is an implementation detail and has been removed